Add Conan package info to visualstudio,msbuild generators

thorntonryan opened the pull request:
Add `CONAN_PACKAGE_NAME` and `CONAN_PACKAGE_VERSION` to the property file generated by the `visual_studio` and `msbuild` generators, similar to what you see in the CMake generator.

Fixes #7177

Changelog: Feature: Add Conan package name and version to Visual Studio generator properties file.
Docs: Omit

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

thorntonryan committed:
Add Conan package info to visualstudio generator

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=7645) <br/>All committers have signed the CLA.

memsharded commented on conans/client/generators/visualstudio.py:
```diff
@@ -116,4 +117,29 @@ def content(self):
             formatted_template = re.sub(userprofile, "$(USERPROFILE)", formatted_template,
                                         flags=re.I)
+
+        if self.conanfile.name or self.conanfile.version:
```
I think maybe it is better to add this to the template instead of adding them dynamically. 

I am fine if they are an empty string if not defined, it is typically easier to check an empty string than a missing variable.

Also, the casing of the variables better be aligned with the existing ones:
```xml
<ConanCompilerFlags>{compiler_flags}</ConanCompilerFlags>
    <ConanLinkerFlags>{linker_flags}</ConanLinkerFlags>
    <ConanPreprocessorDefinitions>{definitions}</ConanPreprocessorDefinitions>
```

So probably ``ConanPackageName``, ``ConanPackageVersion``

memsharded replied:
Consider adding a test also if possible, to validate the new functionality is there and it is not broken in future changes. Ask for help if you need, or we can also add them later.

thorntonryan replied:
> I think maybe it is better to add this to the template instead of adding them dynamically.

Ok, Agreed. I see a path forward now.

> I am fine if they are an empty string if not defined, it is typically easier to check an empty string than a missing variable.

That was my hesitancy. And I didn't really want to tackle turning this into a Jinga template -- *not sure why that isn't being used here*.  The `visualstudio_multi` generator was using `minidom` so I figured I'd roll the dice ... but was sort of expecting it to get shot down :)

memsharded replied:
We try to use the templates whenever possible, and only rely on dom modification when it is really necessary, like the "multi" functionality that needs to insert new nodes, one per install. Maybe not completely necessary to use jinja2 template, maybe the current string format is enough for this one.

thorntonryan replied:
> Consider adding a test also if possible, to validate the new functionality is there and it is not broken in future changes. Ask for help if you need, or we can also add them later.

I've added a test ... but am struggling to get the test suite to run. Need to ask @puetzk when he gets in.

thorntonryan replied:
Looking at the `msbuild` generator, I'm guessing I'd want to add `ConanPackageName` and `ConanPackageVersion` here:
https://github.com/conan-io/conan/blob/0437d7a516213b011f70d27fd314652d8fca8f43/conans/client/generators/msbuild.py#L13-L19

But I'm not yet sure where's the appropriate spot to perform the substitution.

thorntonryan replied:
Think I have a working `msbuild` integration in ~~https://github.com/conan-io/conan/pull/7645/commits/1f998f478e7a9da812169112d3b145cba590077b~~ https://github.com/conan-io/conan/pull/7645/commits/d8da998a2a73f04f287d74209500ca87a2837e09

memsharded reviewed (commented):
Please have a look to the comments.

I would also strongly encourage to check and start using the ``msbuild`` generator instead. It is more powerful and flexible than this one, and it is very likely that it will become the only one in Conan 2.0.

memsharded commented:
Thanks very much for contributing to Conan. Please check the CLA above, it needs to be signed.

thorntonryan commented:
> Thanks very much for contributing to Conan. Please check the CLA above, it needs to be signed.

Will do. Waiting on legal since the CLAs aren't technically shared between conan repos.
_The CLA for cmake-conan was approved. I expect this one to be approved also._

memsharded commented on conans/test/functional/generators/visual_studio_test.py:
```diff
@@ -72,6 +72,8 @@ def package_info(self):
             from conans import ConanFile, MSBuild
             import os
             class HelloConan(ConanFile):
+                name = "Hello Conan"
```
Spaces are not allowed in a package name, I would say this should fail.

thorntonryan replied:
That is fully consistent with me not running the tests! I apologize.

I'm on Windows. And running `tox` at the project root yields:
```bat
C:\projects\conan>python -mvenv venv
...
C:\projects\conan>.\venv\Scripts\install tox
...
C:\projects\conan>.\venv\Scripts\pip install -e .
...
C:\projects\conan>.\venv\Scripts\tox
GLOB sdist-make: C:\projects\conan\setup.py
full inst-nodeps: C:\projects\conan\.tox\.tmp\package\1\conan-1.30.0.dev0.zip
full installed: beautifulsoup4==4.9.1,bottle==0.12.18,certifi==2020.6.20,chardet==3.0.4,colorama==0.4.3,conan @ file:///C:/projects/conan/.tox/.tmp/package/1/conan-1.30.0.dev0.zip,deprecation==2.0.7,distro==1.5.0,fasteners==0.15,future==0.18.2,idna==2.10,Jinja2==2.11.2,MarkupSafe==1.1.1,mock==1.3.0,monotonic==1.5,node-semver==0.6.1,nose==1.3.7,packaging==20.4,parameterized==0.7.4,patch-ng==1.17.4,pbr==5.5.0,pluginbase==1.0.0,Pygments==2.6.1,PyJWT==1.7.1,pyparsing==2.4.7,python-dateutil==2.8.1,PyYAML==5.3.1,requests==2.24.0,six==1.15.0,soupsieve==2.0.1,tqdm==4.48.2,urllib3==1.25.10,waitress==1.4.4,WebOb==1.8.6,WebTest==2.0.35
full run-test-pre: PYTHONHASHSEED='196'
full run-test: commands[0] | nosetests

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
_______________________________________________________ summary _______________________________________________________
  full: commands succeeded
  congratulations :)
```
Which isn't very helpful. I'm on Python 3.8.4 ... and `tox` runs for a coworker who is on 3.7.2 ... figures :)

thorntonryan replied:
Having more luck w/ Python 3.7.9, it appears to be doing something at least. Unsure if it's a Python 3.8 ism or just a "I'm bad about using venvs and boogered up my Python 3.8 install".

thorntonryan replied:
Note to future self, I couldn't get `tox` to find my compiler on Windows, so I had to follow these instructions:
https://github.com/conan-io/conan#without-tox

```sh
$ python -mvenv venv
$ ./venv/Scripts/pip install -e .
$ ./venv/Scripts/python -m pip install -r conans/requirements.txt
$ ./venv/Scripts/python -m pip install -r conans/requirements_server.txt
$ ./venv/Scripts/python -m pip install -r conans/requirements_dev.txt
```

```sh
$ export CONAN_COMPILER="Visual Studio"
$ export CONAN_COMPILER_VERSION=16
```

And use `nose` instead.
```sh
$ ./venv/Scripts/python.exe -m nose -l DEBUG -v conans/test/functional/generators/visual_studio_test.py
```

thorntonryan commented:
Still waiting on the CLA. Converting to draft for now. Will be out of office for a few weeks. Hopefully will have my answer by then. Let me know if you'd rather I close and re-open the PR.

thorntonryan commented:
@memsharded , I think the test cases are working now.

_I haven't tested the full test suite (I don't have the right environment), but hacking the affected tests to run on my machine (i.e. VS2019 w/ no gcc) seems to work._

memsharded commented on conans/client/generators/msbuild.py:
```diff
@@ -87,6 +87,10 @@ def _deps_props(self, name_general, deps):
         template = textwrap.dedent("""\
             <?xml version="1.0" encoding="utf-8"?>
             <Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+                <PropertyGroup Label="Conan-PackageInfo">
```
I am reviewing this, and while it seems correct, I might be missing something.

All other variables are appended with the package name like ``includedirs_pkgname``, because otherwise different packages will overwrite each other values.

The generated .props files will include transitively other .props files if there are transitive dependencies, which is the common case. In this case the ``ConanPackageName`` and ``ConanPackageVersion`` properties will be overwritten, and only 1 value accesible.

It would be necessary to have a complete case, in which the consumption, usage pattern is shown. Why and how a consumer will need this data? Wouldn't it need the data from all transitive dependencies that are being included?

I would say yes, in that case, the solution is a bit more complicated, and we would need at least some cumulative variable or something similar. 

Please tell me what you think.

thorntonryan replied:
James, thank you for your input. I think you raise a valid concern.

My goal is to recover something analogous to:
https://github.com/conan-io/conan/blob/4e3df5fc30bdb63936ab8f196ab8fb56068c511a/conans/client/generators/cmake.py#L94-L95

From the `CMake` generator, which generates a single:
```cmake
### Definition of global aggregated variables ###

set(CONAN_PACKAGE_NAME MyPkg)
set(CONAN_PACKAGE_VERSION 0.1)
```

In the top level `conanbuildinfo.cmake`

I could be mistaken, but I don't think the `CMake` generator worries about the transitive dependencies either.

My goal was to generate `ConanPackageName` and `ConanPackageVersion` only once in the top level property file.

thorntonryan replied:
Our real use case, is not unsurprisingly, a ~~little~~ lot more complicated.

We have a single top level `conanfile.py` that captures the shared build/runtime requirements for a variety of tools/utilities used for the development of embedded software -- think something akin to [build-essential](https://packages.ubuntu.com/xenial/build-essential) but with a lot of our home grown utilities.

To produce these tools, we have a single top level Visual Studio solution with a variety of sub projects that depend on these build / runtime requirements. Each sub project imports the generated `conanbuildinfo.props`.

One project is for a tool that produces artifacts for troubleshooting / debugging embedded  software. Another project contains a mock/example embedded code that we compile and then produce build artifacts for.

We would like to package the artifacts for our example project with the same version as the high level project that produced them -- that way we can say `PkgBuildEssential@August2020` produced `PkgExampleCode@August2020` and keep the two in sync.

This lets downstream tools consuming these example build artifacts obtain good reference files from the latest "internal build essentials" release.

A good spot for pinning a top level version seemed to be the top level `conanfile.py` for the project. And we could get there if the `visual_studio` and `msbuild` generators behaved more like the `cmake` generator and adding something analogous to `CONAN_PACKAGE_NAME` and `CONAN_PACKAGE_VERSION` into one of the generated property files.

_As @puetzk [noted](https://github.com/conan-io/conan/issues/7177#issuecomment-642915060) in the parent issue. We're already taking this approach (i.e. driving top level project information down into sub projects) on other projects built around CMake generators. We were just surprised we couldn't adopt the same approach for .NET projects. Hence the issue and PR :)_

thorntonryan replied:
According to the doc:
> One conan_deps.props Visual Studio properties file, importing all the direct dependencies, in this example both conan_zlib.props and conan_poco.props.

`conan_deps.props` seemed like the top level property file. But a different property file would work just as well if it's a bad fit logically / conceptually.

_In fact, that's how we're currently working around the issue -- by defining a custom generator that produces yet another property file containing this top level information._

memsharded replied:
Oh, I think I understand. It seems that you want the name + version of the current package, not the dependencies, and this is why I was probably confused. Did I understand it right now?

In that case, the place to implement this is most likely the ``toolchain()``, and the ``MSBuildToolchain()``. It is experimental, but it totally sounds like the right place to be defined. Have you had a chance to have a look at them?

thorntonryan replied:
> Oh, I think I understand. It seems that you want the name + version of the current package, not the dependencies, and this is why I was probably confused. Did I understand it right now?

Correct. It sounds like we're on the same page now. I hadn't looked too closely at the new `msbuild` generator and didn't see any references to the toolchain capabilities from the generator page. This must've been what was being [hinted at](https://github.com/conan-io/conan/issues/7177#issuecomment-643137653) earlier.  Sorry I didn't make the connection.

> In that case, the place to implement this is most likely the toolchain(), and the MSBuildToolchain(). It is experimental, but it totally sounds like the right place to be defined. Have you had a chance to have a look at them?

conan_toolchain.props does seem like a better fit. Would you rather I put ` <PropertyGroup Label="Conan-PackageInfo">` there?

memsharded replied:
Hi @thorntonryan

yes, definitely, I suggest moving this name-version info to the generated ``conan_toolchain.props`` from the ``MSBuildToolchain()``, this is the new proposed integration with build systems and where we are putting our efforts.

memsharded reviewed (changes requested):
Hi @thorntonryan 

I have realized that we were missing the transitive dependencies problem, and thus the solution might not be that evident, this would only work for 1st level deps. Please have a look at the comments and let me know.

thorntonryan committed:
Add package info to msbuild toolchain

thorntonryan commented:
@memsharded , I've updated the PR but I have a few observations about `MSBuildToolchain()`:

The [doc](https://docs.conan.io/en/latest/creating_packages/toolchains/msbuild.html) says:
>The MSBuildToolchain will generate two files after a conan install command or before calling the build() method when the package is building in the cache:

But based on my testing, the files are not produced when calling `conan install`.

I added:
```python
from conans import ConanFile, MSBuildToolchain

class BuildToolsConanFile(ConanFile):
    name = "BuildTools"
    version = "2020.07.08"
    settings = "build_type"
    generators = (
        "msbuild"
    )

def toolchain(self):
    tc = MSBuildToolchain(self)
    tc.write_toolchain_files()
```

I think my code is in the right place, but the `toolchain` method never gets called so I'm wondering if there's still something I'm not understanding about how `MSBuildToolchain()` is expected to be used.


And unrelated nit:

From the [msbuild generator](https://docs.conan.io/en/latest/reference/generators/msbuild.html) doc:
> One properties file for each dependency and transitive dependency, like conan_zlib.props, conan_openssl.props*and *conan_poco.props. These files will transitively import other files, in this case as the poco package depends on openssl, the conan_poco.props will import conan_openssl.props file.

We actually have an internal conan package called `toolchain`. If I understand the intent behind `MSBuildToolchain`, then you probably want to guard against creating a package with that reserved name to avoid collisions with the property files produced by `msbuild` generator  for each transitive dependency.

Otherwise I think you can have an opportunity for the `conan_toolchain.props` file to be overwritten by the toolchain file (or vice versa) or even prevent a toolchain file from being written.

memsharded commented:
Hi @thorntonryan 

It seems your toolchain method is non indented correctly, so it is understood as a global, unrelated method, not a recipe one. Try:

```python
class BuildToolsConanFile(ConanFile):
    name = "BuildTools"
    version = "2020.07.08"
    settings = "build_type"
    generators = (
        "msbuild"
    )

    def toolchain(self):
        tc = MSBuildToolchain(self)
        tc.write_toolchain_files()
```

> We actually have an internal conan package called toolchain. If I understand the intent behind MSBuildToolchain, then you probably want to guard against creating a package with that reserved name to avoid collisions with the property files produced by msbuild generator for each transitive dependency.

Haven't thought of that 😮 We might need to think another name, but collision will always be a possibility... lets try to make it at least less likely

thorntonryan commented:
> It seems your toolchain method is non indented correctly, so it is understood as a global, unrelated method, not a recipe one. 

That did it! Ok, I'll work to get unit tests added.

memsharded commented:
Did PR https://github.com/conan-io/conan/pull/8073 to address the conflict name issue.

thorntonryan commented:
> Did PR #8073 to address the conflict name issue.

I think that would work to mitigate against collisions with existing packages.

Even if someone created a package "conantoolchain" and required it, that would result in a transitive dependency property file of  `conan_conantoolchain.props`, which differs from your `conantoolchain.props`.

thorntonryan commented on conans/test/unittests/client/toolchain/test_msbuild.py:
```diff
@@ -0,0 +1,37 @@
+        client.run('install . -s os=Windows -s compiler="Visual Studio" -s compiler.version=15'
+                   ' -s compiler.runtime=MD')
+
+        conan_toolchain_props = client.load("conan_toolchain.props")
```
Given #8073, likely needs rebased/changed to:

```diff
-conan_toolchain_props = client.load("conan_toolchain.props")
+conan_toolchain_props = client.load("conantoolchain.props")
```

memsharded committed:
Merge branch 'develop' into visualstudio_package_info

memsharded committed:
merged develop

memsharded commented:
I merged latest changed from develop, moved the test and fixed a bug in the test (using ``Pk2g`` name), and it passed CI.

It has been merged, it will be in next 1.32 release. Thanks very much for your contribution!
